diff -up firefox-63.0/toolkit/moz.configure.disable-elfhack firefox-63.0/toolkit/moz.configure
--- firefox-63.0/toolkit/moz.configure.disable-elfhack	2018-10-18 13:35:35.870039190 +0200
+++ firefox-63.0/toolkit/moz.configure	2018-10-18 13:36:41.682515492 +0200
@@ -1010,7 +1010,7 @@ with only_when('--enable-compile-environ
                help='Disable elf hacks')
 
         set_config('USE_ELF_HACK',
-                   depends_if('--enable-elf-hack')(lambda _: True))
+                   depends_if('--enable-elf-hack')(lambda _: False))
 
 
 @depends(check_build_environment)
