diff -up firefox-58.0/js/src/jit/none/MacroAssembler-none.h.old firefox-58.0/js/src/jit/none/MacroAssembler-none.h
--- firefox-58.0/js/src/jit/none/MacroAssembler-none.h.old	2018-01-24 20:33:52.569147937 +0100
+++ firefox-58.0/js/src/jit/none/MacroAssembler-none.h	2018-01-24 20:34:06.985088057 +0100
@@ -88,7 +88,7 @@ static constexpr Register WasmTableCallI
 static constexpr Register WasmTlsReg { Registers::invalid_reg };
 
 static constexpr uint32_t ABIStackAlignment = 4;
-static constexpr uint32_t CodeAlignment = 4;
+static constexpr uint32_t CodeAlignment = 8;
 static constexpr uint32_t JitStackAlignment = 8;
 static constexpr uint32_t JitStackValueAlignment = JitStackAlignment / sizeof(Value);
 
