diff -up firefox-102.0/xpcom/idl-parser/xpidl/xpidl.py.python-311 firefox-102.0/xpcom/idl-parser/xpidl/xpidl.py
--- firefox-102.0/xpcom/idl-parser/xpidl/xpidl.py.python-311	2022-07-15 14:23:10.890868155 +0200
+++ firefox-102.0/xpcom/idl-parser/xpidl/xpidl.py	2022-07-15 14:42:17.308836813 +0200
@@ -1572,7 +1572,7 @@ class IDLParser(object):
     t_ignore = " \t"
 
     def t_multilinecomment(self, t):
-        r"/\*(?s).*?\*/"
+        r"/\*(?s:.)*?\*/"
         t.lexer.lineno += t.value.count("\n")
         if t.value.startswith("/**"):
             self._doccomments.append(t.value)
