diff --git a/widget/gtk/ScreenHelperGTK.cpp b/widget/gtk/ScreenHelperGTK.cpp
--- a/widget/gtk/ScreenHelperGTK.cpp
+++ b/widget/gtk/ScreenHelperGTK.cpp
@@ -433,19 +433,21 @@
     return -1;
   }
 
   GdkRectangle workArea;
   if (!GdkMonitorGetWorkarea(monitor, &workArea)) {
+    LOG_SCREEN("  failed, can't get work area");
     return -1;
   }
 
   for (unsigned int i = 0; i < mMonitors.Length(); i++) {
     // Although Gtk/Mutter is very creative in reporting various screens sizes
     // we can rely on Gtk work area start position to match wl_output.
     if (mMonitors[i]->x == workArea.x && mMonitors[i]->y == workArea.y) {
-      LOG_SCREEN(" monitor %d values %d %d -> %d x %d", i, mMonitors[i]->x,
-                 mMonitors[i]->y, mMonitors[i]->width, mMonitors[i]->height);
+      LOG_SCREEN(" monitor %d work area [%d, %d] -> (%d x %d) scale %d", i,
+                 mMonitors[i]->x, mMonitors[i]->y, mMonitors[i]->width,
+                 mMonitors[i]->height, mMonitors[i]->scale);
       return i;
     }
   }
 
   return -1;

