--- firefox-116.0/toolkit/moz.configure	2023-07-27 20:36:17.000000000 +0200
+++ firefox-116.0-telemetry/toolkit/moz.configure	2023-08-03 00:08:46.032857520 +0200
@@ -1043,17 +1042,0 @@
-option(
-    env="MOZ_TELEMETRY_REPORTING",
-    default=mozilla_official,
-    help="Enable telemetry reporting",
-)
-
-set_define("MOZ_TELEMETRY_REPORTING", True, when="MOZ_TELEMETRY_REPORTING")
-
-
-@depends("MOZ_TELEMETRY_REPORTING", milestone.is_nightly)
-def telemetry_on_by_default(reporting, is_nightly):
-    return reporting and is_nightly
-
-
-set_define("MOZ_TELEMETRY_ON_BY_DEFAULT", True, when=telemetry_on_by_default)
-
-
@@ -3073 +3055,0 @@
-    "MOZ_TELEMETRY_REPORTING",
@@ -3078,2 +3060,2 @@
-def data_reporting(telemetry, healthreport, crashreporter, normandy):
-    return telemetry or healthreport or crashreporter or normandy
+def data_reporting(healthreport, crashreporter, normandy):
+    return healthreport or crashreporter or normandy
