diff -U0 firefox-87.0/browser/components/enterprisepolicies/tests/xpcshell/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/browser/components/enterprisepolicies/tests/xpcshell/xpcshell.ini
--- firefox-87.0/browser/components/enterprisepolicies/tests/xpcshell/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:48:28.000000000 +0100
+++ firefox-87.0/browser/components/enterprisepolicies/tests/xpcshell/xpcshell.ini	2021-03-24 08:42:32.727231888 +0100
@@ -9 +8,0 @@
-[test_appupdateurl.js]
@@ -27 +25,0 @@
-[test_sorted_alphabetically.js]
diff -U0 firefox-87.0/devtools/client/shared/remote-debugging/adb/xpcshell/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/devtools/client/shared/remote-debugging/adb/xpcshell/xpcshell.ini
--- firefox-87.0/devtools/client/shared/remote-debugging/adb/xpcshell/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:48:29.000000000 +0100
+++ firefox-87.0/devtools/client/shared/remote-debugging/adb/xpcshell/xpcshell.ini	2021-03-24 08:42:32.727231888 +0100
@@ -9 +8,0 @@
-[test_adb.js]
diff -U0 firefox-87.0/netwerk/test/unit_ipc/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/netwerk/test/unit_ipc/xpcshell.ini
--- firefox-87.0/netwerk/test/unit_ipc/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:49:05.000000000 +0100
+++ firefox-87.0/netwerk/test/unit_ipc/xpcshell.ini	2021-03-24 14:32:31.849567535 +0100
@@ -73 +72,0 @@
-[test_dns_service_wrap.js]
@@ -112 +110,0 @@
-[test_trr_httpssvc_wrap.js]
diff -U0 firefox-87.0/netwerk/test/unit/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/netwerk/test/unit/xpcshell.ini
--- firefox-87.0/netwerk/test/unit/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:49:00.000000000 +0100
+++ firefox-87.0/netwerk/test/unit/xpcshell.ini	2021-03-24 14:30:45.090135351 +0100
@@ -204 +203,0 @@
-[test_dns_service.js]
@@ -228 +226,0 @@
-[test_file_protocol.js]
@@ -336 +333,0 @@
-[test_unix_domain.js]
@@ -348 +344,0 @@
-[test_udp_multicast.js]
@@ -397,2 +392,0 @@
-[test_tls_flags.js]
-skip-if = (verify && (os == 'linux')) || (os == "android" && processor == "x86_64")
@@ -414 +407,0 @@
-[test_network_connectivity_service.js]
@@ -509 +501,0 @@
-[test_httpssvc_retry_with_ech.js]
@@ -519 +510,0 @@
-[test_odoh.js]
diff -U0 firefox-87.0/security/manager/ssl/tests/unit/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/security/manager/ssl/tests/unit/xpcshell.ini
--- firefox-87.0/security/manager/ssl/tests/unit/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:49:06.000000000 +0100
+++ firefox-87.0/security/manager/ssl/tests/unit/xpcshell.ini	2021-03-24 08:42:32.728231864 +0100
@@ -117,2 +116,0 @@
-[test_encrypted_client_hello.js]
-run-sequentially = hardcoded ports
@@ -182 +179,0 @@
-[test_oskeystore.js]
diff -U0 firefox-87.0/security/manager/ssl/tests/unit/xpcshell-smartcards.ini.firefox-tests-xpcshell firefox-87.0/security/manager/ssl/tests/unit/xpcshell-smartcards.ini
--- firefox-87.0/security/manager/ssl/tests/unit/xpcshell-smartcards.ini.firefox-tests-xpcshell	2021-03-18 14:49:06.000000000 +0100
+++ firefox-87.0/security/manager/ssl/tests/unit/xpcshell-smartcards.ini	2021-03-24 08:42:32.728231864 +0100
@@ -10 +9,0 @@
-[test_pkcs11_module.js]
diff -U0 firefox-87.0/toolkit/components/commandlines/test/unit/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/toolkit/components/commandlines/test/unit/xpcshell.ini
--- firefox-87.0/toolkit/components/commandlines/test/unit/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:51:57.000000000 +0100
+++ firefox-87.0/toolkit/components/commandlines/test/unit/xpcshell.ini	2021-03-24 08:42:32.728231864 +0100
@@ -10 +9,0 @@
-[test_resolvefile.js]
diff -U0 firefox-87.0/toolkit/components/extensions/test/xpcshell/xpcshell-common.ini.firefox-tests-xpcshell firefox-87.0/toolkit/components/extensions/test/xpcshell/xpcshell-common.ini
--- firefox-87.0/toolkit/components/extensions/test/xpcshell/xpcshell-common.ini.firefox-tests-xpcshell	2021-03-18 14:51:46.000000000 +0100
+++ firefox-87.0/toolkit/components/extensions/test/xpcshell/xpcshell-common.ini	2021-03-24 08:42:32.728231864 +0100
@@ -79,2 +78,0 @@
-[test_ext_downloads_misc.js]
-skip-if = os == "android" || (os=='linux' && bits==32) || tsan # linux32: bug 1324870, tsan: bug 1612707
@@ -250,2 +247,0 @@
-[test_proxy_listener.js]
-skip-if = appname == "thunderbird"
diff -U0 firefox-87.0/toolkit/modules/subprocess/test/xpcshell/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/toolkit/modules/subprocess/test/xpcshell/xpcshell.ini
--- firefox-87.0/toolkit/modules/subprocess/test/xpcshell/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:51:47.000000000 +0100
+++ firefox-87.0/toolkit/modules/subprocess/test/xpcshell/xpcshell.ini	2021-03-24 08:42:32.728231864 +0100
@@ -10 +9,0 @@
-[test_subprocess.js]
diff -U0 firefox-87.0/toolkit/mozapps/extensions/test/xpcshell/xpcshell.ini.firefox-tests-xpcshell firefox-87.0/toolkit/mozapps/extensions/test/xpcshell/xpcshell.ini
--- firefox-87.0/toolkit/mozapps/extensions/test/xpcshell/xpcshell.ini.firefox-tests-xpcshell	2021-03-18 14:51:57.000000000 +0100
+++ firefox-87.0/toolkit/mozapps/extensions/test/xpcshell/xpcshell.ini	2021-03-24 08:42:32.728231864 +0100
@@ -97 +96,0 @@
-[test_sideloads_after_rebuild.js]
