diff -up firefox-91.0.1/netwerk/protocol/http/nsHttpHandler.cpp.fedora-ua firefox-91.0.1/netwerk/protocol/http/nsHttpHandler.cpp
--- firefox-91.0.1/netwerk/protocol/http/nsHttpHandler.cpp.fedora-ua	2021-08-16 20:10:57.000000000 +0200
+++ firefox-91.0.1/netwerk/protocol/http/nsHttpHandler.cpp	2021-08-23 12:07:21.128772096 +0200
@@ -716,7 +716,7 @@ void nsHttpHandler::BuildUserAgent() {
   // than if we didn't preallocate at all.
   mUserAgent.SetCapacity(mLegacyAppName.Length() + mLegacyAppVersion.Length() +
 #ifndef UA_SPARE_PLATFORM
-                         mPlatform.Length() +
+                         mPlatform.Length() + 10 +
 #endif
                          mOscpu.Length() + mMisc.Length() + mProduct.Length() +
                          mProductSub.Length() + mAppName.Length() +
@@ -735,6 +735,7 @@ void nsHttpHandler::BuildUserAgent() {
   if (!mPlatform.IsEmpty()) {
     mUserAgent += mPlatform;
     mUserAgent.AppendLiteral("; ");
+    mUserAgent.AppendLiteral("Fedora; ");
   }
 #endif
   if (!mCompatDevice.IsEmpty()) {
