diff --git a/widget/gtk/nsWindow.cpp b/widget/gtk/nsWindow.cpp
--- a/widget/gtk/nsWindow.cpp
+++ b/widget/gtk/nsWindow.cpp
@@ -6948,11 +6948,15 @@ nsWindow::GetWaylandDisplay()
   GdkDisplay* gdkDisplay = gdk_display_get_default();
   return mIsX11Display ? nullptr :
                          sGdkWaylandDisplayGetWlDisplay(gdkDisplay);
 }
 
 wl_surface*
 nsWindow::GetWaylandSurface()
 {
-  return moz_container_get_wl_surface(MOZ_CONTAINER(mContainer));
+  if (mContainer)
+    return moz_container_get_wl_surface(MOZ_CONTAINER(mContainer));
+
+  NS_WARNING("nsWindow::GetWaylandSurfaces(): We don't have any mContainer for drawing!");
+  return nullptr;
 }
 #endif

