diff -up mozilla-beta-64737c752ac4af4766ad6f82720818521f3aca24/toolkit/components/remote/nsDBusRemoteService.cpp.1434565 mozilla-beta-64737c752ac4af4766ad6f82720818521f3aca24/toolkit/components/remote/nsDBusRemoteService.cpp
--- mozilla-beta-64737c752ac4af4766ad6f82720818521f3aca24/toolkit/components/remote/nsDBusRemoteService.cpp.1434565	2018-02-01 14:14:38.902873024 +0100
+++ mozilla-beta-64737c752ac4af4766ad6f82720818521f3aca24/toolkit/components/remote/nsDBusRemoteService.cpp	2018-02-01 14:18:02.906281711 +0100
@@ -89,7 +89,11 @@ nsDBusRemoteService::OpenURL(DBusMessage
     errorMsg = nsPrintfCString("org.mozilla.%s.Error", mAppName.get());
     reply = dbus_message_new_error(msg, errorMsg.get(), "Wrong argument");
   } else {
-    nsRemoteService::HandleCommandLine(commandLine, nullptr, 0);
+    guint32 timestamp = gtk_get_current_event_time();
+    if (timestamp == GDK_CURRENT_TIME) {
+        timestamp = guint32(g_get_monotonic_time() / 1000);
+    }
+    nsRemoteService::HandleCommandLine(commandLine, nullptr, timestamp);
     reply = dbus_message_new_method_return(msg);
   }
 
