diff --git a/tools/profiler/rust-api/build.rs b/tools/profiler/rust-api/build.rs
--- a/tools/profiler/rust-api/build.rs
+++ b/tools/profiler/rust-api/build.rs
@@ -61,6 +61,7 @@
     let mut builder = Builder::default()
         .enable_cxx_namespaces()
         .with_codegen_config(CodegenConfig::TYPES | CodegenConfig::VARS | CodegenConfig::FUNCTIONS)
+        .disable_untagged_union()
         .size_t_is_usize(true);
 
     for dir in SEARCH_PATHS.iter() {

