diff -up firefox-69.0/widget/gtk/WindowSurfaceWayland.cpp.old firefox-69.0/widget/gtk/WindowSurfaceWayland.cpp
--- firefox-69.0/widget/gtk/WindowSurfaceWayland.cpp.old	2019-09-11 14:34:36.687382704 +0200
+++ firefox-69.0/widget/gtk/WindowSurfaceWayland.cpp	2019-09-11 14:39:02.000391578 +0200
@@ -751,6 +751,10 @@ static bool IsPopupFullScreenUpdate(Layo
 already_AddRefed<gfx::DrawTarget> WindowSurfaceWayland::Lock(
     const LayoutDeviceIntRegion& aRegion) {
   MOZ_ASSERT(mIsMainThread == NS_IsMainThread());
+  
+  // Disable all commits from frame callback handler and delayed comit handler
+  // as we're updated by gecko compositor.
+  mPendingCommit = false;
 
   LayoutDeviceIntRect lockedScreenRect = mWindow->GetBounds();
   gfx::IntRect bounds = aRegion.GetBounds().ToUnknownRect();
