diff -up firefox-109.0/third_party/libwebrtc/moz.build.ppc-mobzuild firefox-109.0/third_party/libwebrtc/moz.build
--- firefox-109.0/third_party/libwebrtc/moz.build.ppc-mobzuild	2023-01-12 21:02:15.000000000 +0100
+++ firefox-109.0/third_party/libwebrtc/moz.build	2023-01-16 13:30:28.404450100 +0100
@@ -621,7 +621,9 @@ if CONFIG["CPU_ARCH"] == "ppc64" and CON
         "/third_party/libwebrtc/api/audio_codecs/isac/audio_decoder_isac_float_gn",
         "/third_party/libwebrtc/api/audio_codecs/isac/audio_encoder_isac_float_gn",
         "/third_party/libwebrtc/modules/audio_coding/isac_c_gn",
-        "/third_party/libwebrtc/modules/audio_coding/isac_gn"
+        "/third_party/libwebrtc/modules/audio_coding/isac_gn",
+        "/third_party/libwebrtc/modules/desktop_capture/desktop_capture_gn",
+        "/third_party/libwebrtc/modules/desktop_capture/primitives_gn"
     ]
 
 if CONFIG["CPU_ARCH"] == "x86" and CONFIG["OS_TARGET"] == "Linux":
