diff --git a/memory/build/Utils.h b/memory/build/Utils.h
--- a/memory/build/Utils.h
+++ b/memory/build/Utils.h
@@ -67,19 +67,19 @@ constexpr size_t operator"" _MiB(unsigne
   return size_t(aNum) * 1024_KiB;
 }
 
 constexpr size_t operator"" _MiB(long double aNum)
 {
   return size_t(aNum * 1024_KiB);
 }
 
-constexpr long double operator""_percent(long double aPercent)
+constexpr double operator""_percent(long double aPercent)
 {
-  return aPercent / 100;
+  return double(aPercent) / 100;
 }
 
 // Helper for (fast) comparison of fractions without involving divisions or
 // floats.
 class Fraction
 {
 public:
   explicit constexpr Fraction(size_t aNumerator, size_t aDenominator)
diff --git a/memory/build/mozjemalloc.cpp b/memory/build/mozjemalloc.cpp
--- a/memory/build/mozjemalloc.cpp
+++ b/memory/build/mozjemalloc.cpp
@@ -908,18 +908,18 @@ struct arena_bin_t
 
   // Offset of first region in a run for this bin's size class.
   uint32_t mRunFirstRegionOffset;
 
   // Current number of runs in this bin, full or otherwise.
   unsigned long mNumRuns;
 
   // Amount of overhead runs are allowed to have.
-  static constexpr long double kRunOverhead = 1.6_percent;
-  static constexpr long double kRunRelaxedOverhead = 2.4_percent;
+  static constexpr double kRunOverhead = 1.6_percent;
+  static constexpr double kRunRelaxedOverhead = 2.4_percent;
 
   // Initialize a bin for the given size class.
   // The generated run sizes, for a page size of 4 KiB, are:
   //   size|run       size|run       size|run       size|run
   //  class|size     class|size     class|size     class|size
   //     4   4 KiB      8   4 KiB     16   4 KiB     32   4 KiB
   //    48   4 KiB     64   4 KiB     80   4 KiB     96   4 KiB
   //   112   4 KiB    128   8 KiB    144   4 KiB    160   8 KiB

