diff -up firefox-57.0b5/widget/gtk/mozgtk/mozgtk.c.1405267 firefox-57.0b5/widget/gtk/mozgtk/mozgtk.c
--- firefox-57.0b5/widget/gtk/mozgtk/mozgtk.c.1405267	2017-09-19 06:18:28.000000000 +0200
+++ firefox-57.0b5/widget/gtk/mozgtk/mozgtk.c	2017-10-06 12:29:46.941289468 +0200
@@ -95,6 +95,7 @@ STUB(gdk_window_get_update_area)
 STUB(gdk_window_get_user_data)
 STUB(gdk_window_get_visual)
 STUB(gdk_window_get_width)
+STUB(gdk_window_get_window_type)
 STUB(gdk_window_hide)
 STUB(gdk_window_input_shape_combine_region)
 STUB(gdk_window_invalidate_rect)
@@ -520,7 +521,6 @@ STUB(gdk_error_trap_pop_ignored)
 STUB(gdk_event_get_source_device)
 STUB(gdk_screen_get_monitor_workarea)
 STUB(gdk_window_get_type)
-STUB(gdk_window_get_window_type)
 STUB(gdk_x11_window_get_xid)
 STUB(gdk_x11_display_get_type)
 STUB(gdk_wayland_display_get_type)
