diff --git a/widget/xremoteclient/DBusRemoteClient.cpp b/widget/xremoteclient/DBusRemoteClient.cpp
--- a/widget/xremoteclient/DBusRemoteClient.cpp
+++ b/widget/xremoteclient/DBusRemoteClient.cpp
@@ -137,17 +137,17 @@ DBusRemoteClient::GetRemoteDestinationNa
                                                             profileName.get());
     if (aDestinationName.Length() > DBUS_MAXIMUM_NAME_LENGTH)
       aDestinationName.Truncate(DBUS_MAXIMUM_NAME_LENGTH);
 
     static auto sDBusValidateBusName =
       (bool (*)(const char *, DBusError *))
       dlsym(RTLD_DEFAULT, "dbus_validate_bus_name");
     if (!sDBusValidateBusName) {
-      return false
+      return false;
     }
 
     if (!sDBusValidateBusName(aDestinationName.get(), nullptr)) {
       // We don't have a valid busName yet - try to create a default one.
       aDestinationName = nsPrintfCString("org.mozilla.%s.%s", aProgram,
                                                              "default");
       if (!sDBusValidateBusName(aDestinationName.get(), nullptr)) {
         // We failed completelly to get a valid bus name - just quit

