diff -up firefox-121.0.1/toolkit/moz.configure.wasi firefox-121.0.1/toolkit/moz.configure
--- firefox-121.0.1/toolkit/moz.configure.wasi	2024-02-01 09:14:33.816548952 +0100
+++ firefox-121.0.1/toolkit/moz.configure	2024-02-01 09:15:53.264684309 +0100
@@ -2663,7 +2663,7 @@ with only_when(requires_wasm_sandboxing
     def wasi_sysroot_flags(wasi_sysroot):
         if wasi_sysroot:
             log.info("Using wasi sysroot in %s", wasi_sysroot)
-            return ["--sysroot=%s" % wasi_sysroot]
+            return ["--sysroot=%s" % wasi_sysroot, "-nodefaultlibs",  "-lc", "-lwasi-emulated-process-clocks", "-lc++", "-lc++abi", "/raid/CVS/firefox/firefox-127.0.2/wasi-sdk-20/build/compiler-rt/lib/wasi/libclang_rt.builtins-wasm32.a"]
         return []
 
     set_config("WASI_SYSROOT", wasi_sysroot)
diff -up firefox-121.0.1/wasi-sdk-20/Makefile.fedora firefox-121.0.1/wasi-sdk-20/Makefile
--- firefox-121.0.1/wasi-sdk-20/Makefile.fedora	2024-03-04 13:20:56.903403496 +0100
+++ firefox-121.0.1/wasi-sdk-20/Makefile	2024-03-04 13:19:59.381140045 +0100
@@ -44,7 +44,7 @@ endif
 
 # Only the major version is needed for Clang, see https://reviews.llvm.org/D125860.
 CLANG_VERSION=$(shell $(BASH) ./llvm_version_major.sh $(LLVM_PROJ_DIR))
-VERSION:=$(shell $(BASH) ./version.sh)
+VERSION:=20 #$(shell $(BASH) ./version.sh)
 DEBUG_PREFIX_MAP=-fdebug-prefix-map=$(ROOT_DIR)=wasisdk://v$(VERSION)
 
 default: build
@@ -262,8 +262,8 @@ package: build/package.BUILT
 
 build/package.BUILT: build strip
 	mkdir -p dist
-	./deb_from_installation.sh $(shell pwd)/dist "$(VERSION)" "$(BUILD_PREFIX)"
-	./tar_from_installation.sh "$(shell pwd)/dist" "$(VERSION)" "$(BUILD_PREFIX)"
+	#./deb_from_installation.sh $(shell pwd)/dist "$(VERSION)" "$(BUILD_PREFIX)"
+	#./tar_from_installation.sh "$(shell pwd)/dist" "$(VERSION)" "$(BUILD_PREFIX)"
 	touch build/package.BUILT
 
 .PHONY: default clean build strip package check
