diff -up firefox-57.0b5/memory/build/mozjemalloc.cpp.old firefox-57.0b5/memory/build/mozjemalloc.cpp
--- firefox-57.0b5/memory/build/mozjemalloc.cpp.old	2017-10-06 12:00:27.938687498 +0200
+++ firefox-57.0b5/memory/build/mozjemalloc.cpp	2017-10-06 12:02:45.232151309 +0200
@@ -780,7 +780,7 @@ struct arena_t {
  * controlling the malloc behavior are defined as compile-time constants
  * for best performance and cannot be altered at runtime.
  */
-#if !defined(__ia64__) && !defined(__sparc__) && !defined(__mips__) && !defined(__aarch64__)
+#if !defined(__ia64__) && !defined(__sparc__) && !defined(__mips__) && !defined(__aarch64__) && !(defined(__powerpc__))
 #define MALLOC_STATIC_SIZES 1
 #endif
 
